public/app/features/explore/TraceView/components/utils/convertTraceToFlameGraph.ts

```diff
@@ -0,0 +1,111 @@
+import { groupBy, map } from 'lodash';
+
+import { DataFrame, MutableDataFrame } from '@grafana/data';
+
+import { Trace, TraceSpan } from '../types';
+
+interface FlameSpan extends TraceSpan {
+  children: FlameSpan[];
+  total: number;
+  self: number;
+}
+
+export function convertTraceToFlameGraph(trace: Trace): DataFrame {
+  const result: { names: string[]; levels: number[]; values: number[]; self: number[] } = {
+    names: [],
+    levels: [],
+    values: [],
+    self: [],
+  };
+
+  // Step 1: converting spans to a tree
+
+  const spans: Record<string, FlameSpan> = {};
+  const root: FlameSpan = { children: [] } as unknown as FlameSpan;
+  (trace.spans as FlameSpan[]).forEach((span) => {
+    span.children = [];
+    spans[span.spanID] = span;
+  });
+
+  (trace.spans as FlameSpan[]).forEach((span) => {
+    let node = root;
+    if (span.references && span.references.length > 0) {
+      node = spans[span.references[0].spanID] || root;
+    }
+
+    node.children.push(span);
+  });
+
+  // Step 2: group spans with same name
+
+  function groupSpans(span: FlameSpan, d: number) {
+    (span.children || []).forEach((x) => groupSpans(x, d + 1));
```

Comment:
There are recursion limits for all the browsers, even though high https://stackoverflow.com/questions/2805172/what-are-the-js-recursion-limits-for-firefox-chrome-safari-ie-etc I think we could hit them with 100k long traces.